caddyconfig/httploader.go

```diff
@@ -119,6 +124,31 @@ func (hl HTTPLoader) LoadConfig(ctx caddy.Context) ([]byte, error) {
 	return result, nil
 }
 
+// Reattempts the http call using exponential back off when maxRetries is greater than 0.
+func doHttpCallWithRetries(client *http.Client, request *http.Request, maxRetries int) (*http.Response, error) {
+	var resp *http.Response
+	var err error
+
+	for retry := 0; retry <= maxRetries; retry++ {
+		if retry > 0 {
+			caddy.Log().Error(err.Error())
+			exponentialBackOff := math.Pow(2, float64(retry))
+			caddy.Log().Sugar().Infof("reattempting http load in %g seconds (%v retries remain)", exponentialBackOff, maxRetries-retry)
+			time.Sleep(time.Duration(exponentialBackOff) * time.Second)
```

Comment:
Oh, fancy. I wasn't expecting exponential backoff. Props for the effort here.

Some thoughts:

Overall the backoff implementation looks sound. How useful is it? Usually, backoff is implemented for shared or public resources where we need to be careful and mindful clients (and is only useful if most clients implement it, too). For something like config loading, do we expect that backing off will be needed? How many clients are slamming the config source? I'm guessing not a lot (like, not tens-to-hundreds of thousands or more)... so I don't know if it has much utility?

Backoff does also allow the developer more time between retries to fix the problem. A few seconds for the first few retries, and by the time you get to the 20th retry, it's already been a few weeks. That might be an unnecessarily long time between retries.

So one question I have is, how important is having the backoff? And if we do keep it, should we cap the max time between retries to something more reasonable? The longer that span gets, the more time there will be between the backend being available and the actual config being obtained.

If it was just a linear loop (no backoff), the amount of time you spend retrying changes from 2^N seconds 2*N seconds where N is the number of retries (assuming 2 seconds between each one). Or, you could instead have the user configure how _long_ to retry for, and you choose how often to retry. Or, you could have the user configure how long _between each retry_, either forever or with a hard cap on when to give up.

I'm not sure which of those options is preferable here, but what are your thoughts?

Also, we try to avoid ever using `time.Sleep()` because if a config change needs to happen sooner, it will either block or leak resources (continue to run in the background, with no way to stop it). Instead:

```go
select {
    case <-time.After(interval):
    case <-ctx.Done():
        return ctx.Err()
}
```

^ that is the general idea; always select on the Context's Done channel so if it gets canceled, you can stop waiting.

(And even using `time.After()` might not be ideal; the godoc recommends creating a `time.NewTimer()` so it can be stopped and cleaned up if needed. But same idea.)

Reply:
Exponential back-offs are probably overkill now that I think about it. I was modeling that off of hashicorp's lib. A linear approach would be better probably. 2 seconds between firing would be fine.

In my use case, I'd want it to fire indefinitely until it works. Using a large number like 100000 for `max_retries` would be sufficient as the problem is apparent within hours. 

As for the `time.Sleep()` replacement, I'm not sure how to implement that properly so I might need some guidance there.

Reply:
Cool; so maybe if you want it to try indefinitely, specify the time between each retry instead.

I can help with the time.Sleep replacement; it's basically what I showed you, just have to hold onto the context you get during provision.